File: speech_to_text/ruby/real_time_transcription/realtime/realtime.rb
require 'json'
require 'websocket-client-simple'

websocket_url = 'wss://stt-rt.soniox.com/transcribe-websocket'
file_to_stream = 'coffee_shop.pcm_s16le'

# Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
api_key = ENV['SONIOX_API_KEY']

# Connect to WebSocket API
puts 'Opening WebSocket connection...'
ws = WebSocket::Client::Simple.connect(websocket_url)

send_audio_thread = nil

ws.on :open do
  # Send start request
  ws.send({
    'api_key' => api_key,
    'audio_format' => 'pcm_s16le',
    'sample_rate' => 16000,
    'num_channels' => 1,
    'model' => 'stt-rt-preview',
    'language_hints' => []
  }.to_json)

  # Start a thread to send audio
  send_audio_thread = Thread.new do
    # Read and send audio data from file over WebSocket connection
    File.open(file_to_stream, 'rb') do |file|
      loop do
        data = file.read(3840)
        break if data.nil?

        ws.send(data, type: :binary)
        # Sleep for 120 ms
        sleep 0.12
      end
    end

    # Signal end of file
    ws.send('')
  end
  puts 'Transcription started.'
end

ws.on :message do |msg|
  if msg.type == :text
    # Receive text
    response = JSON.parse(msg.data)

    if response['error_code']
      puts "\nError: #{response['error_code']} #{response['error_message']}"
      exit 1
    end

    if response['tokens']
      response['tokens'].each do |token|
        print token['text'] if token['text']
      end
    end

    if response['finished']
      puts "\nTranscription done."
      ws.close
    end
  end
end

ws.on :error do |e|
  # handle WebSocket connection errors
  STDERR.puts "Error: #{e.message}"
end

ws.thread.join

send_audio_thread.join unless send_audio_thread.nil?

File: speech_to_text/nodejs/real_time_transcription/realtime/realtime.js
import { createReadStream } from "fs"

const WEBSOCKET_URL = "wss://stt-rt.soniox.com/transcribe-websocket"
const FILE_TO_STREAM = "coffee_shop.pcm_s16le"

// Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
const API_KEY = process.env.SONIOX_API_KEY;

// Connect to WebSocket API
const ws = new WebSocket(WEBSOCKET_URL);

console.log("Opening WebSocket connection...")
ws.addEventListener("open", async () => {
  // Send start request
  ws.send(
    JSON.stringify({
      api_key: API_KEY,
      audio_format: "pcm_s16le",
      sample_rate: 16000,
      num_channels: 1,
      model: "stt-rt-preview"
    }),
  );

  // Read and send audio data from file over WebSocket connection
  const audioStream = createReadStream(FILE_TO_STREAM, {
    highWaterMark: 3840,
  });

  console.log("Transcription started.")
  for await (const chunk of audioStream) {
    ws.send(chunk);
    await new Promise((resolve) => setTimeout(resolve, 120));
  }

  // Signal end of file
  ws.send("");
});

ws.addEventListener("message", (event) => {
  // Receive and process messages
  const res = JSON.parse(event.data);

  if (res.error_code) {
    console.log(`\nError: ${res.error_code} ${res.error_message}`);
    process.exit(1);
  }

  for (const token of res.tokens) {
    if (token.text) {
      process.stdout.write(token.text);
    }
  }

  if (res.finished) {
    console.log("\nTranscription done.");
  }
});

ws.addEventListener("close", () => {});

ws.addEventListener("error", (error) => {
  // handle WebSocket connection errors
  console.error(error.error);
});
